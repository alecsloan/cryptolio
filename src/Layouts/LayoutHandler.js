import React from 'react'
import CardGrid from './CardGrid'
import AssetTable from './AssetTable'
import PortfolioAreaStackChart from './PortfolioAreaStackChart'
import PortfolioDonutChart from './PortfolioDonutChart'

function LayoutHandler (props) {
  let assets = props.assets

  if (props.settings.sorting === 'price') {
    assets = assets.sort((a, b) => b.price - a.price)
  } else if (props.settings.sorting === 'market_cap') {
    assets = assets.sort((a, b) => b.market_cap - a.market_cap)
  } else if (props.settings.sorting === 'percent_change_1h') {
    assets = assets.sort((a, b) => b.percent_change_1h - a.percent_change_1h)
  } else if (props.settings.sorting === 'percent_change_24h') {
    assets = assets.sort((a, b) => b.percent_change_24h - a.percent_change_24h)
  } else if (props.settings.sorting === 'percent_change_7d') {
    assets = assets.sort((a, b) => b.percent_change_7d - a.percent_change_7d)
  } else {
    assets = assets.sort((a, b) => ((b.holdings || 0.000001) * b.price) - ((a.holdings || 0.000001) * a.price))
  }

  switch (props.renderStyle || props.settings.renderStyle) {
    case "card:classic":
    case "card:compact":
      return <CardGrid assets={assets} renderStyle={props.renderStyle || props.settings.renderStyle} {...props} />
    case "portfolio:chart":
      return <PortfolioAreaStackChart assets={assets} {...props} />
    case "portfolio:donut":
      return <PortfolioDonutChart assets={assets} {...props} />
    case "table":
    default:
      return <AssetTable assets={assets} {...props} />
  }
}

export default LayoutHandler
