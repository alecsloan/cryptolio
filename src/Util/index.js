import { Box, Typography } from '@material-ui/core'
import React from 'react'

export function getCurrencySymbol (currency) {
  return (0).toLocaleString(
    window.navigator.language,
    {
      style: 'currency',
      currency: currency,
      maximumFractionDigits: 0,
      minimumFractionDigits: 0
    }
  ).replace(/\d/g, '').trim()
}

export function getLocalizedNumber (number, settings) {
  return getLocalizedPrice(number, settings, 'decimal')
}

export function getLocalizedPrice (price, settings, style = 'currency') {
  if (!price || !settings) {
    return ''
  }

  let maxDigits = 0

  if (settings.decimals4 && price < settings.decimals4) {
    maxDigits = 4
  } else if (settings.decimals3 && price < settings.decimals3) {
    maxDigits = 3
  } else if (settings.decimals2 && price < settings.decimals2) {
    maxDigits = 2
  }

  return price.toLocaleString(
    window.navigator.language,
    {
      currency: settings.currency,
      maximumFractionDigits: maxDigits,
      minimumFractionDigits: maxDigits,
      style: style
    }
  )
}

export function getLocalizedPercent (number, minimumFractionDigits = 2) {
  return number.toLocaleString(window.navigator.language, { style: 'percent', minimumFractionDigits: minimumFractionDigits })
}

export function MyBalance (props) {
  if (props.holdings > 0 && props.settings.showAssetBalances) {
    return (
      <Typography component='div' id='balance'>
        <Box className='d-inline-block me-1' fontWeight='fontWeightBold'>
          Balance:
        </Box>
        <Box className='d-inline-block pl-2' fontWeight='fontWeightBold'>
          {getLocalizedPrice(props.price * props.holdings, props.settings)}
        </Box>
      </Typography>
    )
  }

  return (
    <div />
  )
}
