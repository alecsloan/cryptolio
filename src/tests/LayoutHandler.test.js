import React from 'react'
import * as Util from '../Util/index'
import LayoutHandler from '../Layouts/LayoutHandler'
import { render, screen } from '@testing-library/react'

const asset = {
  "circulating_supply":222295208,
  "cmcId":1817,
  "holdings":"100",
  "imageURL":"https://s2.coinmarketcap.com/static/img/coins/128x128/1817.png",
  "market_cap":561192867.4385666,
  "max_supply":222295209,
  "name":"Voyager Token",
  "percent_change_1h":0.44361563,
  "percent_change_24h":-1.44690034,
  "percent_change_7d":9.05001049,
  "price":2.52453875406332,
  "rank":108,
  "symbol":"VGX",
  "total_supply":222295208.238,
  "url":"https://coinmarketcap.com/currencies/voyager-token",
  "volume_24h":2506623.63393105
}

let settings = {
  balanceChangeTimeframe: 'percent_change_24h',
  currency: 'USD',
  decimals3: 3,
  show1hChange: true,
  show24hChange: true,
  show7dChange: true,
  showAssetBalances: true,
  showPortfolioBalance: true,
  theme: {
    palette: {
      text: {
        primary: 'black'
      }
    }
  }
}

function testCard(div, type) {
  expect(div.querySelector('img').getAttribute("alt")).toBe(asset.name + ' Logo')
  expect(div.querySelector('img').getAttribute("src")).toBe(asset.imageURL)
  expect(div.querySelector('h5').textContent).toBe(`${asset.name} (${asset.symbol})`)
  expect(div.querySelector("[id='1h-change']").textContent).toBe("1h" + Util.getLocalizedPercent(asset.percent_change_1h * .01))
  expect(div.querySelector("[id='24h-change']").textContent).toBe("24h" + Util.getLocalizedPercent(asset.percent_change_24h * .01))
  expect(div.querySelector("[id='7d-change']").textContent).toBe("7d" + Util.getLocalizedPercent(asset.percent_change_7d * .01))
  expect(div.querySelector("[id='balance']").textContent).toBe('Balance:' + Util.getLocalizedPrice(asset.holdings * asset.price, settings))

  let price = Util.getLocalizedPrice(asset.price, settings)
  let marketCap = "$561.19M"

  if (type === 'classic') {
    price = "Price:" + price
    marketCap = "Market Cap:" + marketCap
  }

  expect(div.querySelector("[id='price']").textContent).toBe(price)
  expect(div.querySelector("[id='market-cap']").textContent).toBe(marketCap)
}

function testTable(div) {
  const row = div.querySelector('[data-id="VGX"]')

  expect(row.querySelector('img').getAttribute("alt")).toBe(asset.name + ' Logo')
  expect(row.querySelector('img').getAttribute("src")).toBe(asset.imageURL)
  expect(row.querySelector("[data-field='id']").textContent).toBe(asset.symbol)
  expect(row.querySelector("[data-field='name']").textContent).toBe(asset.name)
  expect(row.querySelector("[data-field='balance']").textContent).toBe(Util.getLocalizedPrice(asset.holdings * asset.price, settings))
  expect(row.querySelector("[data-field='holdings']").textContent).toBe(Util.getLocalizedNumber(asset.holdings, settings))
  expect(row.querySelector("[data-field='price']").textContent).toBe(Util.getLocalizedPrice(asset.price, settings))
  expect(row.querySelector("[data-field='percent_change_1h']").textContent).toBe(Util.getLocalizedPercent(asset.percent_change_1h * .01))
  expect(row.querySelector("[data-field='percent_change_24h']").textContent).toBe(Util.getLocalizedPercent(asset.percent_change_24h * .01))
  expect(row.querySelector("[data-field='percent_change_7d']").textContent).toBe(Util.getLocalizedPercent(asset.percent_change_7d * .01))
  expect(row.querySelector("[data-field='market_cap']").textContent).toBe('$561.19M')
  expect(row.querySelector("[data-field='volume_24h']").textContent).toBe('$2.51M')
  expect(row.querySelector("[data-field='supply']").textContent).toBe('100%')
}

test('Compact AssetCard renders via the LayoutHandler', async () => {
  render(
    <LayoutHandler
      assets={[asset]}
      editSetting={() => {}}
      removeCrypto={() => {}}
      renderStyle={"card:classic"}
      settings={settings}
      setAssetPanelShown={() => {}}
    />
  )

  const cards = document.getElementsByClassName('MuiCard-root')

  expect(cards.length).toBe(1)

  expect(cards[0]).toMatchSnapshot()

  testCard(cards[0], 'classic')
})

test('Compact AssetCard renders via the LayoutHandler', async () => {
  render(
    <LayoutHandler
      assets={[asset]}
      editSetting={() => {}}
      removeCrypto={() => {}}
      renderStyle={"card:compact"}
      settings={settings}
      setAssetPanelShown={() => {}}
    />
  )

  const cards = document.getElementsByClassName('MuiCard-root')

  expect(cards.length).toBe(1)

  expect(cards[0]).toMatchSnapshot()

  testCard(cards[0], 'compact')
})

test('Portfolio Area Stack Chart renders via the LayoutHandler', async () => {
  render(
    <LayoutHandler
      assets={[asset]}
      editSetting={() => {}}
      removeCrypto={() => {}}
      renderStyle={"portfolio:chart"}
      settings={settings}
      setAssetPanelShown={() => {}}
    />
  )

  expect(document.querySelector("canvas")).toMatchSnapshot()

  const cards = document.getElementsByClassName('MuiCard-root')

  expect(cards.length).toBe(1)

  testCard(cards[0], 'classic')
})

test('Portfolio Area Stack Chart with Compact Card sub style renders via the LayoutHandler', async () => {
  settings.renderSubStyle = "card:compact";

  render(
    <LayoutHandler
      assets={[asset]}
      editSetting={() => {}}
      removeCrypto={() => {}}
      renderStyle={"portfolio:chart"}
      settings={settings}
      setAssetPanelShown={() => {}}
    />
  )

  expect(document.querySelector("canvas")).toMatchSnapshot()

  const cards = document.getElementsByClassName('MuiCard-root')

  expect(cards.length).toBe(1)

  testCard(cards[0], 'compact')
})

test('Portfolio Area Stack Chart with Table sub style renders via the LayoutHandler', async () => {
  settings.renderSubStyle = "table";

  render(
    <LayoutHandler
      assets={[asset]}
      editSetting={() => {}}
      removeCrypto={() => {}}
      renderStyle={"portfolio:chart"}
      settings={settings}
      setAssetPanelShown={() => {}}
    />
  )

  expect(document.querySelector("canvas")).toMatchSnapshot()

  expect(document.querySelectorAll("[data-id]").length).toBe(1)

  testTable(document)
})

test('Portfolio Donut Chart renders via the LayoutHandler', async () => {
  render(
    <LayoutHandler
      assets={[asset]}
      editSetting={() => {}}
      removeCrypto={() => {}}
      renderStyle={"portfolio:donut"}
      settings={settings}
      setAssetPanelShown={() => {}}
    />
  )

  expect(document.querySelector("canvas")).toMatchSnapshot()

  expect(document.querySelectorAll("[data-id]").length).toBe(1)

  testTable(document)
})

test('Portfolio Donut Chart with Classic Card sub style renders via the LayoutHandler', async () => {
  settings.renderSubStyle = "card:classic";

  render(
    <LayoutHandler
      assets={[asset]}
      editSetting={() => {}}
      removeCrypto={() => {}}
      renderStyle={"portfolio:donut"}
      settings={settings}
      setAssetPanelShown={() => {}}
    />
  )

  expect(document.querySelector("canvas")).toMatchSnapshot()

  const cards = document.getElementsByClassName('MuiCard-root')

  expect(cards.length).toBe(1)

  testCard(cards[0], 'classic')
})

test('Portfolio Donut Chart with Compact Card sub style renders via the LayoutHandler', async () => {
  settings.renderSubStyle = "card:compact";

  render(
    <LayoutHandler
      assets={[asset]}
      editSetting={() => {}}
      removeCrypto={() => {}}
      renderStyle={"portfolio:donut"}
      settings={settings}
      setAssetPanelShown={() => {}}
    />
  )

  expect(document.querySelector("canvas")).toMatchSnapshot()

  const cards = document.getElementsByClassName('MuiCard-root')

  expect(cards.length).toBe(1)

  testCard(cards[0], 'compact')
})

test('Asset Table renders via the LayoutHandler', async () => {
  render(
    <LayoutHandler
      assets={[asset]}
      editSetting={() => {}}
      renderStyle={"table"}
      settings={settings}
      setAssetPanelShown={() => {}}
    />
  )

  expect(document.querySelectorAll("[data-id]").length).toBe(1)

  expect(document.querySelectorAll("[data-id]").length).toBe(1)

  testTable(document)
})

test('Mobile Asset Table renders via the LayoutHandler', async () => {
  global.innerWidth = 500;

  render(
    <LayoutHandler
      assets={[asset]}
      editSetting={() => {}}
      renderStyle={"table"}
      settings={settings}
      setAssetPanelShown={() => {}}
    />
  )

  const intervalSelector = document.querySelector('[aria-label="percent change"]')

  expect(intervalSelector).toMatchSnapshot()

  expect(document.querySelectorAll("[data-id]").length).toBe(1)

  const row = document.querySelector('[data-id="VGX"]')

  expect(row.querySelector('img').getAttribute("alt")).toBe(asset.name + ' Logo')
  expect(row.querySelector('img').getAttribute("src")).toBe(asset.imageURL)
  expect(screen.getByText(asset.rank))
  expect(screen.getByText(asset.symbol))
  expect(screen.getByText(asset.name))
  expect(screen.getByText(Util.getLocalizedPercent(asset.percent_change_24h * .01)))
  expect(row.querySelector("[data-field='price']").textContent).toBe(Util.getLocalizedPrice(asset.price, settings))
  expect(screen.getByText(Util.getLocalizedPrice(asset.holdings * asset.price, settings)))
  expect(screen.getByText(Util.getLocalizedNumber(asset.holdings, settings) + " " + asset.symbol))
})