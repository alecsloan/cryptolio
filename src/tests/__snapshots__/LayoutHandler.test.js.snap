// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compact AssetCard renders via the LayoutHandler 1`] = `
<div
  class="MuiPaper-root MuiCard-root card MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    class="MuiCardHeader-root h-0 p-0"
  >
    <div
      class="MuiCardHeader-content"
    />
    <div
      class="MuiCardHeader-action m-0 p-2"
    >
      <button
        aria-label="Voyager Token settings"
        class="MuiButtonBase-root MuiIconButton-root p-0 settings MuiIconButton-colorInherit"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiCardContent-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <img
          alt="Voyager Token Logo"
          height="100"
          src="https://s2.coinmarketcap.com/static/img/coins/128x128/1817.png"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5"
        >
          Voyager Token (VGX)
        </h5>
      </div>
    </div>
  </div>
  <div
    class="MuiCardContent-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiTypography-root MuiTypography-body1"
              id="1h-change"
            >
              <div
                class="MuiBox-root MuiBox-root-1 d-inline-block me-2"
              >
                1h
              </div>
              <div
                class="MuiBox-root MuiBox-root-2 d-inline-block pl-2"
              >
                <span>
                  0.44%
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiTypography-root MuiTypography-body1"
              id="24h-change"
            >
              <div
                class="MuiBox-root MuiBox-root-3 d-inline-block me-2"
              >
                24h
              </div>
              <div
                class="MuiBox-root MuiBox-root-4 d-inline-block pl-2"
              >
                <span>
                  -1.45%
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiTypography-root MuiTypography-body1"
              id="7d-change"
            >
              <div
                class="MuiBox-root MuiBox-root-5 d-inline-block me-2"
              >
                7d
              </div>
              <div
                class="MuiBox-root MuiBox-root-6 d-inline-block pl-2"
              >
                <span>
                  9.05%
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        <div>
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
              id="price"
            >
              <div
                class="MuiTypography-root MuiTypography-body1"
                id="undefined-change"
              >
                <div
                  class="MuiBox-root MuiBox-root-7 d-inline-block me-1"
                >
                  Price:
                </div>
                <div
                  class="MuiBox-root MuiBox-root-8 d-inline-block pl-2"
                >
                  $2.525
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
              id="market-cap"
            >
              <div
                class="MuiTypography-root MuiTypography-body1"
                id="undefined-change"
              >
                <div
                  class="MuiBox-root MuiBox-root-9 d-inline-block me-1"
                >
                  Market Cap:
                </div>
                <div
                  class="MuiBox-root MuiBox-root-10 d-inline-block pl-2"
                >
                  $561.19M
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                class="MuiTypography-root MuiTypography-body1"
                id="balance"
              >
                <div
                  class="MuiBox-root MuiBox-root-11 d-inline-block me-1"
                >
                  Balance:
                </div>
                <div
                  class="MuiBox-root MuiBox-root-12 d-inline-block pl-2"
                >
                  $252
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Compact AssetCard renders via the LayoutHandler 2`] = `
<div
  class="MuiPaper-root MuiCard-root card MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    class="MuiCardHeader-root"
  >
    <div
      class="MuiCardHeader-avatar"
    >
      <img
        alt="Voyager Token Logo"
        height="70"
        src="https://s2.coinmarketcap.com/static/img/coins/128x128/1817.png"
      />
    </div>
    <div
      class="MuiCardHeader-content"
    >
      <span
        class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
      >
        <h5
          class="MuiBox-root MuiBox-root-13"
        >
          Voyager Token (VGX)
        </h5>
      </span>
      <span
        class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
      >
        <div>
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
              id="price"
            >
              $2.525
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
              id="market-cap"
            >
              $561.19M
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                class="MuiTypography-root MuiTypography-body1"
                id="balance"
              >
                <div
                  class="MuiBox-root MuiBox-root-14 d-inline-block me-1"
                >
                  Balance:
                </div>
                <div
                  class="MuiBox-root MuiBox-root-15 d-inline-block pl-2"
                >
                  $252
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <div
                class="MuiTypography-root d-inline-block m-1 MuiTypography-body1"
                id="1h-change"
              >
                <div
                  class="MuiBox-root MuiBox-root-16"
                >
                  1h
                </div>
                <div
                  class="MuiBox-root MuiBox-root-17 d-inline-block pl-2"
                >
                  <span>
                    0.44%
                  </span>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <div
                class="MuiTypography-root d-inline-block m-1 MuiTypography-body1"
                id="24h-change"
              >
                <div
                  class="MuiBox-root MuiBox-root-18"
                >
                  24h
                </div>
                <div
                  class="MuiBox-root MuiBox-root-19 d-inline-block pl-2"
                >
                  <span>
                    -1.45%
                  </span>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <div
                class="MuiTypography-root d-inline-block m-1 MuiTypography-body1"
                id="7d-change"
              >
                <div
                  class="MuiBox-root MuiBox-root-20"
                >
                  7d
                </div>
                <div
                  class="MuiBox-root MuiBox-root-21 d-inline-block pl-2"
                >
                  <span>
                    9.05%
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </span>
    </div>
    <div
      class="MuiCardHeader-action"
    >
      <button
        aria-label="Voyager Token settings"
        class="MuiButtonBase-root MuiIconButton-root p-0 settings MuiIconButton-colorInherit"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiCardContent-root"
    hidden=""
  >
    <div
      class="MuiGrid-root MuiGrid-container"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiTypography-root d-inline-block m-1 MuiTypography-body1"
              id="1h-change"
            >
              <div
                class="MuiBox-root MuiBox-root-22"
              >
                1h
              </div>
              <div
                class="MuiBox-root MuiBox-root-23 d-inline-block pl-2"
              >
                <span>
                  0.44%
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiTypography-root d-inline-block m-1 MuiTypography-body1"
              id="24h-change"
            >
              <div
                class="MuiBox-root MuiBox-root-24"
              >
                24h
              </div>
              <div
                class="MuiBox-root MuiBox-root-25 d-inline-block pl-2"
              >
                <span>
                  -1.45%
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiTypography-root d-inline-block m-1 MuiTypography-body1"
              id="7d-change"
            >
              <div
                class="MuiBox-root MuiBox-root-26"
              >
                7d
              </div>
              <div
                class="MuiBox-root MuiBox-root-27 d-inline-block pl-2"
              >
                <span>
                  9.05%
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        <div>
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
              id="price"
            >
              $2.525
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
              id="market-cap"
            >
              $561.19M
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                class="MuiTypography-root MuiTypography-body1"
                id="balance"
              >
                <div
                  class="MuiBox-root MuiBox-root-28 d-inline-block me-1"
                >
                  Balance:
                </div>
                <div
                  class="MuiBox-root MuiBox-root-29 d-inline-block pl-2"
                >
                  $252
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <div
                class="MuiTypography-root d-inline-block m-1 MuiTypography-body1"
                id="1h-change"
              >
                <div
                  class="MuiBox-root MuiBox-root-30"
                >
                  1h
                </div>
                <div
                  class="MuiBox-root MuiBox-root-31 d-inline-block pl-2"
                >
                  <span>
                    0.44%
                  </span>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <div
                class="MuiTypography-root d-inline-block m-1 MuiTypography-body1"
                id="24h-change"
              >
                <div
                  class="MuiBox-root MuiBox-root-32"
                >
                  24h
                </div>
                <div
                  class="MuiBox-root MuiBox-root-33 d-inline-block pl-2"
                >
                  <span>
                    -1.45%
                  </span>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <div
                class="MuiTypography-root d-inline-block m-1 MuiTypography-body1"
                id="7d-change"
              >
                <div
                  class="MuiBox-root MuiBox-root-34"
                >
                  7d
                </div>
                <div
                  class="MuiBox-root MuiBox-root-35 d-inline-block pl-2"
                >
                  <span>
                    9.05%
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Mobile Asset Table renders via the LayoutHandler 1`] = `
<div
  aria-label="percent change"
  class="MuiToggleButtonGroup-root mb-2"
  role="group"
>
  <button
    aria-label="1h percent change"
    aria-pressed="false"
    class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal"
    tabindex="0"
    type="button"
    value="percent_change_1h"
  >
    <span
      class="MuiToggleButton-label"
    >
      1h
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    aria-label="24h percent change"
    aria-pressed="true"
    class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal Mui-selected"
    tabindex="0"
    type="button"
    value="percent_change_24h"
  >
    <span
      class="MuiToggleButton-label"
    >
      24h
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    aria-label="7d percent change"
    aria-pressed="false"
    class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal"
    tabindex="0"
    type="button"
    value="percent_change_7d"
  >
    <span
      class="MuiToggleButton-label"
    >
      7d
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`Portfolio Area Stack Chart renders via the LayoutHandler 1`] = `
<canvas
  data-zr-dom-id="zr_0"
  height="100"
  style="position: absolute; left: 0px; top: 0px; width: 0px; height: 100px; webkit-user-select: none; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0); padding: 0px; margin: 0px; border-width: 0px;"
  width="0"
/>
`;

exports[`Portfolio Area Stack Chart with Compact Card sub style renders via the LayoutHandler 1`] = `
<canvas
  data-zr-dom-id="zr_0"
  height="100"
  style="position: absolute; left: 0px; top: 0px; width: 0px; height: 100px; webkit-user-select: none; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0); padding: 0px; margin: 0px; border-width: 0px;"
  width="0"
/>
`;

exports[`Portfolio Area Stack Chart with Table sub style renders via the LayoutHandler 1`] = `
<canvas
  data-zr-dom-id="zr_0"
  height="100"
  style="position: absolute; left: 0px; top: 0px; width: 0px; height: 100px; webkit-user-select: none; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0); padding: 0px; margin: 0px; border-width: 0px;"
  width="0"
/>
`;

exports[`Portfolio Donut Chart renders via the LayoutHandler 1`] = `
<canvas
  data-zr-dom-id="zr_0"
  height="100"
  style="position: absolute; left: 0px; top: 0px; width: 0px; height: 100px; webkit-user-select: none; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0); padding: 0px; margin: 0px; border-width: 0px;"
  width="0"
/>
`;

exports[`Portfolio Donut Chart with Classic Card sub style renders via the LayoutHandler 1`] = `
<canvas
  data-zr-dom-id="zr_0"
  height="100"
  style="position: absolute; left: 0px; top: 0px; width: 0px; height: 100px; webkit-user-select: none; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0); padding: 0px; margin: 0px; border-width: 0px;"
  width="0"
/>
`;

exports[`Portfolio Donut Chart with Compact Card sub style renders via the LayoutHandler 1`] = `
<canvas
  data-zr-dom-id="zr_0"
  height="100"
  style="position: absolute; left: 0px; top: 0px; width: 0px; height: 100px; webkit-user-select: none; user-select: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0); padding: 0px; margin: 0px; border-width: 0px;"
  width="0"
/>
`;
